### DISCUSSION TITLE
How to spee up simulation with OKane2022 parameters

### DISCUSSION THREAD

**Discussion by lappemic** [pybamm-team/PyBaMM#4307] [Q&A]
I'm working on battery simulations using PyBaMM, and I've run into a performance issue that I'm hoping someone might have some insights on. I'm simulating a lithium-ion battery using the OKane2022 parameter set, and I've noticed that the simulation is taking an extremely long time to complete, especially for the drive cycles.

Here's a brief overview of my setup:
- Using a DFN model with various degradation mechanisms enabled (SEI growth, lithium plating, particle mechanics, etc.)
- Simulating 1000 cycles in total, with 50 cycles per batch
- Using the WLTC drive cycle, scaled by a factor of 0.35
- Resolution set to 10 seconds

I've successfully used a similar setup with the Chen2020 parameters, and it runs in a reasonable amount of time. However, when I switch to the OKane2022 parameters, the simulation slows down (barely simulating 40 cycles in 24h), particularly during the drive cycles.

Has anyone else experienced this issue with the OKane2022 parameters? I'm wondering if there might be some parameter-specific optimizations or solver settings that could help speed things up. Any suggestions for improving the simulation speed while maintaining accuracy would be greatly appreciated.

Some questions I have:
1. Are there any known issues with the OKane2022 parameters that might cause slower simulations?
2. Is there a recommended solver or solver settings for this type of simulation?
3. Are there any other tricks or techniques for speeding up simulations with complex drive cycles?

The code i use:
```python
import pybamm
from src.utils import extract_resolution_str, count_drive_cycles_in_experiment, get_model_acronym

os.chdir(pybamm.__path__[0] + "/..")

# Define simulation parameters
TOTAL_CYCLES = 1000
CYCLES_PER_BATCH = 50
SAVE_EVERY_N_BATCH = 1
RESOLUTION = '(10 seconds period)'
SCALE_FACTOR = 0.35
PARAMETER_REFERENCE = "OKane2022"

parameter_values = pybamm.ParameterValues(PARAMETER_REFERENCE)
parameter_values.update({"SEI kinetic rate constant [m.s-1]": 1e-12})
# parameter_values.set_initial_stoichiometries(1);
var_pts = {
    "x_n": 5,  # negative electrode
    "x_s": 5,  # separator
    "x_p": 5,  # positive electrode
    "r_n": 30,  # negative particle
    "r_p": 30,  # positive particle
}


model = pybamm.lithium_ion.DFN(
    {
        "SEI": "solvent-diffusion limited",
        "SEI porosity change": "true",
        "lithium plating": "partially reversible",
        "lithium plating porosity change": "true",  # alias for "SEI porosity change"
        "particle mechanics": ("swelling and cracking", "swelling only"),
        "SEI on cracks": "true",
        "loss of active material": "stress-driven",
        "calculate discharge energy": "true",  # for compatibility with older PyBaMM versions
    }
)

drive_cycle = pd.read_csv("pybamm/input/drive_cycles/WLTC.csv", comment='#', header=None).to_numpy()

# scale the current of the drive cycle
drive_cycle[:, 1] *= SCALE_FACTOR

# Initialize batch counter to decide when to save intermediate results
batch_counter = 0

for batch in range(0, TOTAL_CYCLES, CYCLES_PER_BATCH):
    # set up the experiment for this batch
    experiment = pybamm.Experiment(
        ([
            # this part is for the first cycle only
            (
                f"Hold at 4.2 V until C/20 {RESOLUTION}",
                f"Rest for 4 hours {RESOLUTION}",
                f"Discharge at 0.3C until 2.5 V {RESOLUTION}",  # initial capacity check
                f"Rest for 30 minutes {RESOLUTION}",
                f"Charge at 0.3C until 4.2 V {RESOLUTION}",
                f"Hold at 4.2 V until C/20 {RESOLUTION}",
                f"Rest for 30 minutes {RESOLUTION}",
            )
        ] if batch == 0 else [])
        + [
            (
                pybamm.step.current(drive_cycle),
                f"Rest for 30 minutes {RESOLUTION}",
                pybamm.step.current(drive_cycle),
                f"Rest for 30 minutes {RESOLUTION}",
                pybamm.step.current(drive_cycle),
                f"Rest for 30 minutes {RESOLUTION}",
                f"Charge at 1C until 4.2V {RESOLUTION}",
                f"Hold at 4.2V until C/20 {RESOLUTION}",
                f"Rest for 30 minutes {RESOLUTION}"
            ),
        ] * min(CYCLES_PER_BATCH, TOTAL_CYCLES - batch)
        + [
            (
                f"Discharge at 0.3C until 2.5 V {RESOLUTION}",
                f"Rest for 30 minutes {RESOLUTION}",
                f"Charge at 0.3C until 4.2 V {RESOLUTION}",
                f"Hold at 4.2 V until C/20 {RESOLUTION}",
                f"Rest for 30 minutes {RESOLUTION}"
            )
        ]
    )

    # trying with another solver - for drive cycles pybamm recommends using the fast solver
    solver = pybamm.CasadiSolver(mode="fast")
    sim = pybamm.Simulation(
        model, 
        parameter_values=parameter_values, 
        experiment=experiment,
        var_pts=var_pts
        ) # , solver=solver) # this is not working for some reason

    try: # this is every batch except the first
        solution = sim.solve(starting_solution=solution, showprogress=True)
    except NameError: # this is the first batch
        solution = sim.solve(showprogress=True)


    # Update the batch counter
    batch_counter += 1

    # Check if the results should be saved
    if batch_counter % SAVE_EVERY_N_BATCH == 0 or batch + CYCLES_PER_BATCH >= TOTAL_CYCLES:
        outfilename_pickle_red = f"{MODELNAME}_{PARAMETER_REFERENCE.lower()}_{count_drive_cycles_in_experiment(experiment)}WLTC_{batch + CYCLES_PER_BATCH:05}_{SCALE_FACTOR_STR}_{extract_resolution_str(RESOLUTION)}_red.pkl"
        outfilepath_pickle_red = OUTDIR.joinpath(outfilename_pickle_red)
        solution.save_data(outfilepath_pickle_red, OUTPUT_VARIABLES, to_format="pickle")
```

**Comment by MarcBerliner** [comment]
Hi @lappemic,

The reason you cannot use `pybamm.CasadiSolver(mode="fast")` is that your experiments include stop conditions like `until 4.2 V` and `until C/20` – fast mode only works with experiments that have a specified final time.

We are actively working on improving the solvers in PyBaMM. The `develop` branch of PyBaMM has an updated version of `pybamm.IDAKLUSolver()`, which should give better performance and reliability over the default Casadi solver. Can you please the IDAKLU solver and see if that improves your performance? Platform-specific installation instructions are here: https://docs.pybamm.org/en/latest/source/user_guide/installation/index.html

**Reply by lappemic** [reply]
Hey @MarcBerliner thank you very much for your feedback! I will try the develp branch on Wednesday and come back to you! Thanks a lot again already again!

**Reply by lappemic** [reply]
Thank you for your guidance on using the `pybamm.IDAKLUSolver()` from the `develop` branch of PyBaMM. I've followed your advice and made the necessary changes to my setup. Here's what I've done:

1. Uninstalled the stable version of PyBaMM.
2. Installed the development branch of PyBaMM using `pip install git+https://github.com/pybamm-team/PyBaMM.git@develop`.
3. Updated my code to use the `IDAKLUSolver` instead of `CasadiSolver(mode="fast")`.

I tried to run this setup then but got the following error
```bash
Traceback (most recent call last):
  File "/home/mlappert/batterypipeline/src/notebooks/pybamm/cell_simulation_wltc_okane2022.py", line 157, in <module>
    solver = pybamm.IDAKLUSolver()
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/mlappert/batterypipeline/venv/lib/python3.11/site-packages/pybamm/solvers/idaklu_solver.py", line 224, in __init__
    raise ImportError("KLU is not installed")
ImportError: KLU is not installed
```

Do you have an idea why this happens?

**Reply by MarcBerliner** [reply]
Hi @lappemic, the installation subpages have a few more steps for installing the optional solvers based on your OS:
- [Windows Subsystem for Linux](https://docs.pybamm.org/en/stable/source/user_guide/installation/windows-wsl.html)
- [GNU Linux and macOS](https://docs.pybamm.org/en/stable/source/user_guide/installation/install-from-source.html)

**Reply by lappemic** [reply]
Sorry, i seem not to be able to figure how to install these. Could you guide me a bit more on this? I am running on an Ubuntu vm.

**Reply by MarcBerliner** [reply]
With the latest 24.9 PyBaMM release, the `IDAKLUSolver` is now available in the default installation. Can you please install pybamm with `pip install pybamm` and trying this again?

**Comment by benshang1997** [comment]
Hi， @lappemic ，I meet this issue recently， the degradation spped is very slow and the cassadi solver may meet some trouble after several cycles, Have you found other better ways? thank you

**Reply by lappemic** [reply]
Hey @benshang1997 i unfortunately was not yet able to check it out. But i plan to try above suggestion from Marc.
